### DISCUSSION TITLE
Any difference between gpio-keys and gpio-leds? Should have consolidate to gpio-signal?

### DISCUSSION THREAD

**Discussion by jettr** [zephyrproject-rtos/zephyr#29281] [General]
I've noticed there are device tree bindings for both [gpio-keys](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/gpio/gpio-keys.yaml) and [gpio-leds](https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/gpio/gpio-leds.yaml) that are identical expect for their description. Should we consolidate these two definitions?

Also I would like to have a list of named gpios that don't fall into the key or led category; therefore, I would like to create a more generic version: `gpio-signals`. The `gpio-signals` binding would be exactly the same as goth `gpio-keys` and `gpio-leds`.

If we have the more generically named version (`gpio-signals`), should we deprecate and remove the `gpio-keys` and `gpio-leds` compatibility bindings and move everything to the `gpio-signals` binding?

----

Related question, but I can create a new discussion if needed: I also noticed that the `GPIO_`* flags indented for [use in the device tree](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/dt-bindings/gpio/gpio.h) only include the [lower 8 bits](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/gpio.h#L297). This does not include whether the gpio is an input or an [output](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/gpio.h#L46). When listing our named gpios (with above `gpio-signals`), it make sense to define in the `dts`  file whether it is an output and what the initial output state should be.

Specifying only the lower 8 bits is not a limit of the device tree, per se, but it looks to be convention. Can or should we change this convention, or what is the motivation behind this convention?

**Comment by galak** [comment]
The `gpio-keys` and `gpio-leds` are bindings we take from linux and thus want to maintain them even if functionally we don't have any difference beyond one being an input and the other and output.

I've avoid accepting a generic gpio signal binding as usually a GPIO should be associated with some other dts node.

We added support for inferred bindings as a middle ground:

https://docs.zephyrproject.org/latest/guides/dts/bindings.html#inferred-bindings

**Reply by jettr** [reply]
I did see the inferred bindings, but I would rather keep all of the named-gpios together under one node. It felt a little weird to just put everything under `zephyr,user`. I really want a label for each GPIO signal as well.

I also agree that each subsystem or device should express which gpios they need in device tree. I imagined that a given gpio pin would be in both the global named-gpio device (for which I want to use the general `gpio-signal`) and also in the subsystem that claims it (like board power sequencing). The subsystem that claims the named gpio could do so by alias to keep everything tidy and single sourced.

I do think there is still room for a list of all of the gpios on the system and human-readable names they map to on the schematics. This is helpful for debugging, especially for a shell command to list all of the gpio levels on the system with their net name.

**Comment by galak** [comment]
> Related question, but I can create a new discussion if needed: I also noticed that the `GPIO_`* flags indented for [use in the device tree](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/dt-bindings/gpio/gpio.h) only include the [lower 8 bits](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/gpio.h#L297). This does not include whether the gpio is an input or an [output](https://github.com/zephyrproject-rtos/zephyr/blob/master/include/drivers/gpio.h#L46). When listing our named gpios (with above `gpio-signals`), it make sense to define in the `dts` file whether it is an output and what the initial output state should be.
> 
> Specifying only the lower 8 bits is not a limit of the device tree, per se, but it looks to be convention. Can or should we change this convention, or what is the motivation behind this convention?

Our general principle is to conform to upstream dts binding conventions as much as possible.  So we utilize concepts from:

https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/tree/Bindings/gpio/gpio.txt

We don't support all the things defined here, but if what you need is covered here we can implement it.

**Reply by jettr** [reply]
I didn't see the `gpio-leds` or `gpio-keys` concept in the linux gpio.txt file. Maybe I missed it?

The example they gave about naming GPIOs were at the gpio controller level, but for Zephyr the gpio controller is defined at the chip level. The names and uses for each pin should really be designated at the board or application level. Does it not feel a little weird to add named gpios to the gpio controller, or maybe I am extracting examples that aren't intended for Zephyr?
